src/core/infer_request.cc

```diff
@@ -181,49 +275,50 @@ InferenceRequest::Normalize(const InferenceBackend& backend)
   for (auto& pr : inputs_) {
     const ModelInput* input_config;
     RETURN_IF_ERROR(backend.GetInput(pr.first, &input_config));
+    auto& input = pr.second;
 
-    auto& shape = *pr.second.MutableShape();
+    auto new_shape = input.MutableShape();
+    *new_shape = input.OriginalShape();
 
     // If the inference request specifies a shape for an input, make
     // sure it matches what the model expects.
-    if (shape.size() > 0) {
-      if (!CompareDimsWithWildcard(input_config->dims(), shape)) {
+    if (new_shape->size() > 0) {
+      if (!CompareDimsWithWildcard(input_config->dims(), *new_shape)) {
         return Status(
             RequestStatusCode::INVALID_ARG,
             "unexpected shape for input '" + pr.first + "' for model '" +
                 model_name_ + "'. Expected " +
                 DimsListToString(input_config->dims()) + ", got " +
-                DimsListToString(shape));
+                DimsListToString(*new_shape));
       }
 
-      // If there is a reshape for this input then clear the dims and
-      // set them to the reshape. As reshape may have variable-size
+      // If there is a reshape for this input then set the shape to
+      // match the reshape. As reshape may have variable-size
       // dimensions, we need to record corresponding value so that we
       // can set the value correctly for reshape.
       if (input_config->has_reshape()) {
         std::deque<int64_t> variable_size_values;
         for (int64_t idx = 0; idx < input_config->dims_size(); idx++) {
           if (input_config->dims(idx) == -1) {
-            variable_size_values.push_back(shape[idx]);
+            variable_size_values.push_back((*new_shape)[idx]);
           }
         }
 
-        shape.clear();
```

Comment:
This is needed, below is appending the reshaped value to the `new_shape`, so it should be empty to start with.

Reply:
Good catch!